fedbiomed/researcher/experiment.py

```diff
@@ -1589,6 +1643,48 @@ def run_once(self, increase: bool = False, test_after: bool = False) -> int:
 
         return 1
 
+    def _run_agg_optimizer(
+        self,
+        aggregated_params: Dict[str, T],
+    ) -> Dict[str, T]:
+        """Optionally refine aggregated parameters into model updates.
+
+        Args:
+            aggregated_params: `Aggregator`-output model weights, equal
+                to $$\\theta^t - aggregate(\\{\\theta_i^{t+1}\\}_{i=1}^I)$$.
+
+        Returns:
+            Updated model weights that should be used to replace the current
+            `self._global_model`. If a researcher optimizer is set, they are
+            obtained by taking a SGD(-based) step over the aggregated updates.
+            Otherwise, the outputs are the same as the inputs.
+        """
+        # If not Optimizer is set, return the inputs.
+        if self._agg_optimizer is None:
+            return aggregated_params
+        # Run any start-of-round routine.
+        self._agg_optimizer.init_round()
+        # Recover the aggregated model updates, wrapped as a Vector.
+        # Optionally restrict weights that require updating to non-frozen ones.
+        # aggregated_params = agg({w^t - sum_k(eta_{k,i,t} * grad_{k,i,t})}_i)
+        # hence aggregated_params = w^t - agg(updates_i)
+        # hence agg_gradients = agg_i(updates_i)
+        names = set(
+            self._job.training_plan.get_model_params(only_trainable=True)
+        )
+        init_params = Vector.build(
+            {k: v for k, v in self._global_model.items() if k in names}
+        )
+        agg_gradients = init_params - Vector.build(
+            {k: v for k, v in aggregated_params.items() if k in names}
+        )
+        # Take an Optimizer step to compute the updates.
+        # When using vanilla SGD: agg_updates = - lrate * agg_gradients
```

Comment:
Nice addition! But I think this could also be tested to make sure non-trainable layers are not updated during the optimization. I can take care of that if you want

Reply:
Great! Thanks for the test, I think it covered well the use case depicted!